### ISSUE TITLE
Distributed tensorflow - Stuck at "CreateSession still waiting for response from worker"

### ISSUE THREAD

**Issue by nolanliou** [tensorflow/tensorflow#12745]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:
binary
- **TensorFlow version (use command below)**:
1.0 && 1.3
- **Python version**: 
2.7
- **CUDA/cuDNN version**:
no

### Describe the problem
Run the code on k8s with the spec `1 ps + 20 workers`. there are some workers print the log all the time.
```
I tensor flow/core/distributed_runtime/master.cc:193] CreateSession still waiting for response from worker: /job:worker/replica:0/task:12
I tensor flow/core/distributed_runtime/master.cc:193] CreateSession still waiting for response from worker: /job:worker/replica:0/task:17
```
I met two situations:
1. worker 1 wait for worker 12 and 17, but worker 12 and 17 could start training without these logs.
2. all of other works wait for the worker 12 and 17.

I could telnet the no-response worker within the docker vm in both situations, so it's wired. Is there a bug? 
Some have also encountered this problem. see [Stackoverflow](https://stackoverflow.com/search?q=CreateSession+still+waiting+for+response+from+worker).
### Source code / logs
```
import datetime
import json
import numpy as np
import os
import tensorflow as tf

flags = tf.app.flags
flags.DEFINE_integer("max_epochs", 10000000, "Number of steps to run trainer.")
flags.DEFINE_string("checkpoint_path", "./checkpoint/",
                    "The checkpoint directory")
flags.DEFINE_string("output_path", "./tensorboard/",
                    "indicates training output")
flags.DEFINE_integer("checkpoint_period", 1,
                     "Number of epochs to save checkpoint.")
flags.DEFINE_float("learning_rate", 0.01, "Initial learning rate.")
FLAGS = flags.FLAGS


def main():
    # Create train data
    train_X = np.linspace(-1, 1, 100)
    train_Y = 2 * train_X + np.random.randn(*train_X.shape) * 0.33 + 10
    learning_rate = FLAGS.learning_rate
    start_training_time = datetime.datetime.now()

    optimizer = tf.train.GradientDescentOptimizer(learning_rate)
    # Exampmle: {"cluster": {"ps": ["127.0.0.1:3001"], "worker": ["127.0.0.1:3002", "127.0.0.1:3003"]}, "task": {"index": 0, "type": "worker"}}
    env = json.loads(os.environ.get("TF_CONFIG", "{}"))
    task_data = env.get("task", None)
    cluster_spec = env["cluster"]
    task_type = task_data["type"]
    task_index = task_data["index"]

    cluster = tf.train.ClusterSpec(cluster_spec)
    server = tf.train.Server(cluster,
                             job_name=task_type,
                             task_index=task_index)

    if task_type == "ps":
        server.join()
    elif task_type == "worker":
        with tf.device(tf.train.replica_device_setter(
                worker_device="/job:{}/task:{}".format(task_type, task_index),
                cluster=cluster)):

            # Define the model
            keys_placeholder = tf.placeholder(tf.int32, shape=[None, 1])
            keys = tf.identity(keys_placeholder)
            X = tf.placeholder("float", shape=[None, 1])
            Y = tf.placeholder("float", shape=[None, 1])
            w = tf.Variable(0.0, name="weight")
            b = tf.Variable(0.0, name="bias")
            global_step = tf.Variable(0, name="global_step", trainable=False)
            loss = tf.reduce_sum(tf.square(Y - tf.multiply(X, w) - b))
            train_op = optimizer.minimize(loss, global_step=global_step)
            predict_op = tf.multiply(X, w) + b
            tf.summary.scalar("loss", loss)
            summary_op = tf.summary.merge_all()
            init_op = tf.global_variables_initializer()
            saver = tf.train.Saver()
            #saver = tf.train.Saver(sharded=True)


            sv = tf.train.Supervisor(is_chief=(task_index == 0),
                                     logdir=FLAGS.checkpoint_path,
                                     init_op=init_op,
                                     #summary_op=summary_op,
                                     summary_op=None,
                                     saver=saver,
                                     global_step=global_step,
                                     save_model_secs=60)

            try:
                with sv.managed_session(server.target) as sess:
                    print("Save tensorboard files into: {}".format(FLAGS.output_path))
                    writer = tf.summary.FileWriter(FLAGS.output_path, sess.graph)

                    print("Run training with epoch number: {}".format(
                        FLAGS.max_epochs))
                    for i in range(FLAGS.max_epochs):
                        for (x, y) in zip(train_X, train_Y):
                            x = np.array([[x]])
                            y = np.array([[y]])
                            sess.run(train_op, feed_dict={X: x, Y: y})

                        if i % FLAGS.checkpoint_period == 0:
                            x = np.array([[train_X[0]]])
                            y = np.array([[train_Y[0]]])
                            summary_value, loss_value, step = sess.run(
                                [summary_op, loss, global_step],
                                feed_dict={X: x,
                                           Y: y})
                            print("Epoch: {}, loss: {}".format(i, loss_value))
                            if task_index == 0:
                                writer.add_summary(summary_value, step)

                    writer.close()

                    end_training_time = datetime.datetime.now()
                    print("[{}] End of distributed training.".format(
                        end_training_time - start_training_time))


            except Exception as e:
                print(e)


if __name__ == "__main__":
    main()
```

**Comment by nolanliou** [comment]
@mrry Could you help to take a look? thanks.

**Comment by mrry** [comment]
Does the problem consistently affect workers 12 and 17, and not other workers? If so, is there anything unusual about those workers' network configurations?

**Comment by nolanliou** [comment]
@mrry, thanks for your reply. 

Yes, it's consistent between worker 12 and 17, they wait for each other and other workers wait for them. But other workers could connect each other, so it's confused.

All workers use the same configuration. use k8s to schedule the network.

**Comment by mrry** [comment]
This sounds like the problem is with your network configuration, rather than a bug in TensorFlow. In particular, there seems to be something wrong with the configuration for workers 12 and 17... perhaps something else is listening on the same socket? Is there anything in the logs for those workers? You could try running the servers with environment variable `GRPC_VERBOSITY=DEBUG` and see if there is any extra information.

Another possible thing you could try is running with 15 workers (instead of 20) to determine if the VM for worker 12 is always affected (independent of cluster configuration). You could try shuffling the list of network addresses used for the workers, to see if the problem still affects the same hosts. Or you could try connecting directly to the server on worker 12/17 (after setting up the cluster) by creating a `tf.Session(ADDRESS_FOR_WORKER_12_OR_17)` and see if you can run a simple graph.

**Comment by nolanliou** [comment]
Thanks for your detailed answer, I'll have a try.

**Comment by OscarDPan** [comment]
@mrry I have the same problem that one worker got stuck at trying to connect to 3 parameter servers (which are apparently running and connected to all other workers). There are other workers on the same machine as the problematic one, and they are all connected to the ps. Also ports are assign to all servers as expected.

This behavior happens quite often and the worker number, ps number or the machine number in problem are apparently random.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue? Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue? Please update the label and/or status accordingly.

**Comment by mrry** [comment]
Closing this issue due to lack of activity. @nolanliou Hopefully things are working now, but feel free to reopen if you can reproduce with a simpler example. @OscarDPan We upgraded to a version of gRPC that fixes a big deadlock bug (that happened more frequently at higher load) in the latest nightly build (will be in TF 1.5), so you might want to try upgrading to see if that fixes the connectivity problems. If it doesn't, please open a new issue with the relevant details.

**Comment by ccyjava** [comment]
I meet the same problem, after several hours debug, i found that because the order of the cluster_spec is incorrect, in other word, the  task_index miss match with the ps/worker list, after i change the order, it fixed....

**Comment by TerryLMay** [comment]
I meet the same problem, and my ps server and worker server correct match with ps/worker index. but there also have this problem.  
```
CreateSession still waiting for response from /job/ps/task/0
```
i check ps and worker python process. they all run in correct way. 
any idea about this issue？ thanks

**Comment by ebadawy** [comment]
@mrry I've the same issue! posted a question on stack overflow with the code that I've tried https://stackoverflow.com/questions/51219374/tensorflow-distributed-createsession-still-waiting-for-response-from-worker-j

I've just tried to run both ps and worker on the same cluster and yet still getting the same message! any idea how to fix/debug that?

**Comment by zywangnpu2013** [comment]
I've the same issue now.Did you solve it?

**Comment by ebadawy** [comment]
Yes, my problem was kinda silly! I ssh to the remote server using `user@example.com` and that what I was using to define the cluster IP address in Tensorflow, but it turned out the I should be only using `example.com` and the problem was solved after that!

another thing that could cause the problem is that you make sure the task id matches with the cluster IP address, for simplicity try with one ps and one worker that both are on the same machine and see if it works with you.

**Comment by codeflitting** [comment]
I've the same issue

chief restart when worker completed will  cause the problem

any ideas？

**Comment by salemmohammed** [comment]
@mrry I have the following error:
CreateSession still waiting for response from worker: /job:ps/replica:0/task:0
Is there any suggestion for this error?

**Comment by edumotya** [comment]
It's not solved yet. The same issue using runtimeVersion 1.12:

```
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:3 I  worker-replica-6
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:4 I  worker-replica-6
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:5 I  worker-replica-6
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:7 I  worker-replica-6
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:8 I  worker-replica-6
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:9 I  worker-replica-6
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:10 I  worker-replica-6
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:11 I  worker-replica-6
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:12 I  worker-replica-6
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:13 I  worker-replica-6
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:14 I  worker-replica-6
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:15 I  worker-replica-6
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:16 I  worker-replica-6
 worker-replica-6 CreateSession still waiting for response from worker: /job:worker/replica:0/task:17 I  worker-replica-6
```

**Comment by ChChwang** [comment]
> Yes, my problem was kinda silly! I ssh to the remote server using `user@example.com` and that what I was using to define the cluster IP address in Tensorflow, but it turned out the I should be only using `example.com` and the problem was solved after that!
> 
> another thing that could cause the problem is that you make sure the task id matches with the cluster IP address, for simplicity try with one ps and one worker that both are on the same machine and see if it works with you.

there are always 2 manchine no matter 8 workers or 6 works, and the IP of them are not fixed.
have you fixed the issue? thanks a lot.

**Comment by rsk2327** [comment]
> I meet the same problem, after several hours debug, i found that because the order of the cluster_spec is incorrect, in other word, the task_index miss match with the ps/worker list, after i change the order, it fixed....

Could you elaborate on this? I am getting the same error and I was hoping to fix it

**Comment by chansonzhang** [comment]
I met the same problem in kubeflow, due to network problems.
I finally solved this by following steps,
1) fix flannel service, and make sure network between each ps/worker or worker/worker pair is reachable. you can check that by curl <ip>:<port> in each tfjob pod;
2) fix dns prolem, and make make sure curl <tfjob_name>:<port> between each pair tfjob pods also works. you should check that all dns service and pods are work fine and dns ip you pass to kubelet.service is the same as the cluster ip you get by `kubectl get service -n kube-system|grep dns`

**Comment by moonyoung430** [comment]
@chansonzhang 
I use command curl <tfjob_name>:<port> in each pod and it report "curl: (7) Failed to connect to 192.168.2.53 port 80: Connection refused",  How to solve this please?

**Comment by chansonzhang** [comment]
> @chansonzhang
> I use command curl <tfjob_name>: in each pod and it report "curl: (7) Failed to connect to 192.168.2.53 port 80: Connection refused", How to solve this please?

seems like your flannel service is not ok

**Comment by chansonzhang** [comment]
> @chansonzhang
> I use command curl <tfjob_name>: in each pod and it report "curl: (7) Failed to connect to 192.168.2.53 port 80: Connection refused", How to solve this please?

seems like your flannel service is not ok
